man/man1/wfontdemo.1
.TH WFONTDEMO 1 "Version 1, Release 4" "W Window System" "W PROGRAMS"
.SH NAME
wfontdemo \- display all available W fonts
.SH SYNOPSIS
wfontdemo <W font directory>
.SH DESCRIPTION
.I Wfontdemo
prints a short string in all available fonts, in all available styles,
scrolling the window if necessary.
.PP
Note that due to directory sorting done by `ls' the fonts may not appear
in the order you might expect them.
.SH BUGS
can't change font dir
.SH LAST CHANGED
15th. Feb. 1996 by TeSche


wt/CHANGES
++oddie 9/2000:
- fixed listbox/filesel resizing & added minimum size setopt.

++bsittler 7/2000:
- Fixed vt color/reverse and attribute resetting
- Add HTTP/1.1 'Host' header to Wetscape requests
- Other minor changes (netgames, wvt, waudio)

++jcarlson 7/2000:
- WTINYSCREEN changes

++eero, 5/1999:
- Fixed key handling for special keys in Wt/getstring.c & wedit/wedit.c.

++eero, 1/1999:
- Added support for new W special keys so that now on W platforms which
implement special key mappings, you don't need to use Emacs shortcuts
with text widgets (or remap your console cursor keys)...


------ notes for version released with W R1PL0 ----------

ADDITIONAL WIDGETS

- Arrow, a simple button with an arrow drawn into it.

- Dial, a dial widget where you can position a needle into desired
  angle.  This isn't in the W toolkit library itself as then you would
  need to link math library to all of your Wt programs.  You'll find it
  in the warzone game tar-ball.

- Edittext, mainly intended for simple one page stuff, but as this
  spaghetti can do some additional stuff like paragraph formatting,
  there's a real abuse built on this, the WEdit.  As edittext widget is
  large and buggy, it's in the WEdit directory instead of the toolkit.

- Menu, a strip of buttons from which popups pop-up.  Both the popups
  and menus use a structure defined in Wt.h instead of strings that Kay
  was using.  The reason is that I want callbacks to follow
  automatically hierarchically opened popups.

- Popup, hierarchical popups.  Can include subpopups, images, lines and
  in addition items can also be disabled and checked.

- Range, a scrollbar composite with which you can select a value between
  given limits.

- Select, a getstring widget coupled with an arrow openable listbox from
  which items can be selected into a getstring widget.


OTHER CHANGES

- vt can now do everything wterm  can do + user can rebind the colors
  in wtrc.

- Added '.wtconfig' file and made makefiles use the W compilation
  configuration file '.config' in parent directory.

- wt_do_event() doesn't call wt_dispatch_event() anymore with
  EVENT_GADGET/GADGET_EXIT event as those might not have valid window
  pointer.  If you want to check for abnormal (unboud) exit, check the
  wt_return value.  GADGET_EXIT returns a negative number and
  GADGET_CLOSE (in shell widget) a positive number now.

- Added xterm mouse events, reverse mode, fd/bgcolor setting + improved
  double/triple click timing in VT widget.

- Aliases a couple new tags into others in HTML widget and added
  HTML header (ignored) and TITLE parsing.

- Widget event processing functions return now unrecognized or
  non-processed WEVENTs or NULL instead of long.  Assumption is that if
  function returns WEVENT, the widget whose action it was, is still
  around.  If in doubt, use NULL.  Unrecognized key events should be
  returned for focus change though.  (actually it could still have been
  an int, but returning a WEVENT offers a possibility for event mapping
  widgets.)

- Added wt_global structure containing Wt configuration variables and
  made widgets to use it.  wt_init() will now read these variables from
  ~/.wtrc + a list of (variable, value) pairs.  Widget specific
  initialization code in their init() functions may read these with the
  wt_variable() function.

- HTML widget fonts can be configured by above method.

- Added wt_loadfont() function which will use the user specified Wt
  default if widget doesn't specify some font parameter.  It also exits
  program if Wt default fonts can't be loaded, because otherwise
  unavailable fonts may chrash naive programs.

- Changed widgets to reflect the new w_loadfont() syntax (NULL/zero
  values tell Wlib/server to use server defaults).

- Added new WT_FONTSIZE define.  The earlier WT_FONT setopt sets now the
  font *family* name, new one sets the font size.

- Getopting WT_FONT returns WFONT* pointer also for above reason.

- Removed 'fontname' members from widget structures.  Needed information
  and more can be gotten from the 'font' member if needed.

- Fileselector has now separate lists for directory and file names.
  Both of course have an autolocator (type & return in addition to point
  & click).

- Wt_keys.h defines values for some special keys like cursors etc.  Half
  of the defines are Emacs style and half what the special keys itself
  give or are in use on other platforms.

- Changed Wt option defines.  There were a couple overlapping ones.


------- Kay's original notes: ----------------------------------------------

- added some missing close() and truncate the file to zero length before
  writing a new selection in selection.c.

- better selection handling in VT widget.

- made everything resizeable.

- applied patches by eero and fixed some more bugs myself.

- implemented a simple clipboard and changed getstring and vt widget to use it.

- fixed several serious bugs in the html widget.

- wrote VT widget.

- added an option to the icon editor to sync edit field with display
  field (simon).

- added missing special chars to html widget (simon).

- added code to the fileselector to set a default file name.

- lots of changes by eero (button.c, fileselect.c, listbox.c, getstring.c), some
  fixes to his changes by me.

- added missing wt_ungetfocus() to listbox and getstring.

- wrote Tk-like packer widget.

- new fileselect widget and testprogram by eero. Added better filename
  matching (eeros version was a joke, really).

- getstring and listbox updates by eero.

- changed the scrollbar widget's user interface to become more user friendly.
  See scrollbar.doc for details on how to use it. Also changed viewport and
  listbox to use the new interface.

- added listbox_focus(). Otherwise listbox cannot receive keyboard focus...

- big bugfix: when wt_delete() is called on a widget all pending timers must
  be canceled! (affected list, iconedit and getstring).


demos/wfontdemo.c
/*
 * programs/wfontdemo.c, part of W (C) 94-02/96 by Torsten Scherer
 * (TeSche) itschere@techfak.uni-bielefeld.de
 *
 * Shows fonts in the given directory (provided that W server can load them).
 *
 * CHANGES
 * - lots and many, TeSche
 */

#include <stdio.h>
#include <dirent.h>
#include <unistd.h>
#include <Wlib.h>


#define WIDTH 600
#define HEIGHT 350

#define TEXT_STYLES	6

static ushort Style[TEXT_STYLES] = {
  F_NORMAL,
  F_REVERSE,
  F_UNDERLINE,
  F_BOLD,
  F_LIGHT,
  F_ITALIC
};


static WWIN *getwindow(short width, short height)
{
  WSERVER *wserver;
  WWIN *win;

  if (!(wserver = w_init())) {
    return NULL;
  }

  if (!(win = w_create(width, height, W_MOVE|W_TITLE|W_CLOSE|EV_KEYS|EV_MOUSE))) {
    return NULL;
  }

  w_settitle(win, " wfontdemo ");

  if (w_open(win, UNDEF, UNDEF) < 0) {
    w_delete(win);
    return NULL;
  }

  return win;
}


int main(int argc, char *argv[])
{
  WWIN *win;
  DIR *dp;
  ushort styles;
  short idx, size;
  struct dirent *dentry;
  const char *family;
  char msg[80];
  WFONT *font;
  WEVENT *ev;

  if (argc != 2 || argv[1][0] == '-') {
    fprintf(stderr, "usage: %s <font directory>\n", *argv);
    return -1;
  }

  if (!(win = getwindow(WIDTH, HEIGHT))) {
    fprintf(stderr, "error: can't open window\n");
    return -1;
  }

  w_setmode(win, M_CLEAR);

  if ((dp = opendir(argv[1]))) {

    /* skip `.' and `..' */
    readdir(dp);
    readdir(dp);

    while ((dentry = readdir(dp))) {
      family = w_fonttype(dentry->d_name, &size, &styles);
      if ((font = w_loadfont(family, size, styles))) {
	sprintf(msg, "%s: Fontdemo ~123@#", dentry->d_name);
	w_setfont(win, font);

	for (idx = 0; idx < TEXT_STYLES; idx++) {
	  w_settextstyle(win, Style[idx]);
	  w_vscroll(win, 0, font->height, WIDTH, HEIGHT - font->height, 0);
	  w_pbox(win, 0, HEIGHT - font->height - 1, WIDTH, font->height);
	  w_printstring(win, 0, HEIGHT - font->height - 1, msg);
	}
	w_unloadfont(font);

	if ((ev = w_queryevent(NULL, NULL, NULL, 2000))) {
	  if ((ev->type == EVENT_GADGET) &&
	      ((ev->key == GADGET_EXIT) || (ev->key == GADGET_CLOSE))) {
	    return -1;
	  }
	}
      } else {
        fprintf (stderr, "unable to load font: '%s'\n", dentry->d_name);
      }
    }

    closedir(dp);
  }

  w_settitle(win, " wfontdemo - click anywhere to dismiss ");

  w_queryevent(NULL, NULL, NULL, -1);
  w_delete(win);

  return 0;
}
